File: docs/palindromes-in-peg/scaffold_stream_heads.py
"""Clonable stream heads using the explicitly scheduled queue kernel.

A right move consumes a front cell at an operation boundary and requests
three subsequent maintenance instructions if it used a queue. The caller
must schedule those before another public operation on that queue.
"""
from scaffold_circuit import (Value, Ref, PREVIOUS, EMPTY, TRUE, FALSE,
                              conjunction, disjunction, neg, choose)
from scaffold_circuit_structs import Stack, StackPool
from scaffold_queue_registers import QueueRegisters


class StreamBank:
  def __init__(self, circuit, names):
    self.circuit = circuit
    self.cells = StackPool(circuit, {"cells": 1})
    self.queues = QueueRegisters(circuit, tuple(name + ".in" for name in names))
    self.heads = {name: StreamHead(self, name) for name in names}

  def finalize(self):
    for head in self.heads.values():
      head.finalize()
    self.queues.finalize()


class StreamHead:
  def __init__(self, bank, name):
    self.circuit, self.name = bank.circuit, name
    self.focus_key = name + ".focus"
    self.focus = self.circuit.get_ref(PREVIOUS, self.focus_key)
    self.left = Stack(bank.cells, name + ".l", "cells")
    self.right = Stack(bank.cells, name + ".r", "cells")
    self.queue = bank.queues.queues[name + ".in"]

  def can_right(self):
    return disjunction(neg(self.right.empty()), neg(self.queue.empty()))

  def peek_right(self, enabled=TRUE):
    from_stack = neg(self.right.empty())
    a, _ = self.right.peek()
    b, _ = self.queue.s["F"].peek()
    self.circuit.require(neg(self.queue.s["F"].empty()),
                         conjunction(enabled, neg(from_stack), neg(self.queue.empty())))
    return Ref.select(from_stack, a, b)

  def move_right(self, enabled):
    self.circuit.require(self.can_right(), enabled)
    self.left.push(self.focus, enabled=enabled, slot=0)
    from_stack = neg(self.right.empty())
    a, _ = self.right.pop(conjunction(enabled, from_stack))
    maintain = conjunction(enabled, neg(from_stack))
    b = self.queue.pop(maintain)
    self.focus = Ref.select(enabled, Ref.select(from_stack, a, b), self.focus)
    return maintain

  def move_left(self, enabled):
    self.circuit.require(neg(self.left.empty()), enabled)
    self.right.push(self.focus, enabled=enabled, slot=0)
    value, _ = self.left.pop(enabled)
    self.focus = Ref.select(enabled, value, self.focus)

  def follow_arrival(self, cell, enabled):
    """The global end head advances to a newly arrived cell directly."""
    self.left.push(self.focus, enabled=enabled, slot=0)
    self.focus = Ref.select(enabled, cell, self.focus)

  def copy_from(self, other, enabled):
    self.focus = Ref.select(enabled, other.focus, self.focus)
    self.left.copy_from(other.left, enabled)
    self.right.copy_from(other.right, enabled)
    self.queue.copy_from(other.queue, enabled)

  def reset(self, enabled):
    self.focus = Ref.select(enabled, EMPTY, self.focus)
    self.left.clear(enabled)
    self.right.clear(enabled)
    self.queue.clear(enabled)

  def finalize(self):
    self.left.finalize()
    self.right.finalize()
    self.circuit.put_ref(self.focus_key, self.focus)


class PatternTextHead:
  def __init__(self, bank, name):
    self.circuit, self.name = bank.circuit, name
    self.pattern = bank.heads[name + ".pattern"]
    self.text = bank.heads[name + ".text"]
    self.mode_key = name + ".text_mode"
    self.text_mode = self.circuit.get(PREVIOUS, self.mode_key, (False, True), False).eq(True)

  def available(self):
    return choose(self.text_mode, self.text.can_right(), self.pattern.focus.present())

  def read(self, enabled=TRUE):
    target = Ref.select(self.text_mode,
                        self.text.peek_right(conjunction(enabled, self.text_mode)), self.pattern.focus)
    value = self.circuit.get(target, "input", tuple("ab"), "a")
    return Value.select(target.present(), value, Value.constant(None))

  def start(self, snapshot, enabled):
    self.pattern.copy_from(snapshot, enabled)
    self.text.reset(enabled)
    self.text_mode = choose(enabled, neg(snapshot.focus.present()), self.text_mode)

  def move(self, direction, enabled):
    mode = self.text_mode
    if direction == 1:
      self.pattern.move_left(conjunction(enabled, neg(mode)))
      requested = self.text.move_right(conjunction(enabled, mode))
      self.text_mode = disjunction(mode, conjunction(enabled, neg(mode), neg(self.pattern.focus.present())))
      return self.text.queue.name, requested
    if direction == -1:
      boundary = neg(self.text.focus.present())
      self.text.move_left(conjunction(enabled, mode, neg(boundary)))
      requested = self.pattern.move_right(conjunction(enabled, disjunction(neg(mode), boundary)))
      self.text_mode = conjunction(mode, neg(conjunction(enabled, mode, boundary)))
      return self.pattern.queue.name, requested
    raise ValueError("unit direction required")

  def copy_from(self, other, enabled):
    self.pattern.copy_from(other.pattern, enabled)
    self.text.copy_from(other.text, enabled)
    self.text_mode = choose(enabled, other.text_mode, self.text_mode)

  def finalize(self):
    self.circuit.put(self.mode_key, Value.select(self.text_mode, Value.constant(True), Value.constant(False)))


class OrientedHead:
  """A frozen forward or backward view, preserved by head copies."""
  def __init__(self, bank, name):
    self.circuit, self.name = bank.circuit, name
    self.cursor = bank.heads[name]
    self.mode_key = name + ".reverse"
    self.reverse = self.circuit.get(PREVIOUS, self.mode_key, (False, True), False).eq(True)

  def start(self, snapshot, reverse, enabled):
    self.cursor.copy_from(snapshot, enabled)
    self.reverse = choose(enabled, TRUE if reverse else FALSE, self.reverse)

  def read(self, enabled=TRUE):
    forward = self.cursor.peek_right(conjunction(enabled, neg(self.reverse)))
    target = Ref.select(self.reverse, self.cursor.focus, forward)
    value = self.circuit.get(target, "input", tuple("ab"), "a")
    return Value.select(target.present(), value, Value.constant(None))

  def move(self, direction, enabled):
    if direction not in (-1, 1):
      raise ValueError("unit oriented movement required")
    right = conjunction(enabled, neg(self.reverse) if direction == 1 else self.reverse)
    left = conjunction(enabled, self.reverse if direction == 1 else neg(self.reverse))
    requested = self.cursor.move_right(right)
    self.cursor.move_left(left)
    return self.cursor.queue.name, requested

  def copy_from(self, other, enabled):
    self.cursor.copy_from(other.cursor, enabled)
    self.reverse = choose(enabled, other.reverse, self.reverse)

  def finalize(self):
    self.circuit.put(self.mode_key, Value.select(self.reverse, Value.constant(True), Value.constant(False)))


class MirrorHead:
  """A frozen view of u # reverse(u), without building that word."""
  def __init__(self, bank, name):
    self.circuit, self.name = bank.circuit, name
    self.forward = bank.heads[name + ".forward"]
    self.reverse = bank.heads[name + ".reverse"]
    self.phase_key, self.nonempty_key = name + ".view_phase", name + ".nonempty"
    self.phase = self.circuit.get(PREVIOUS, self.phase_key, ("forward", "separator", "reverse", "end"), "separator")
    self.nonempty = self.circuit.get(PREVIOUS, self.nonempty_key, (False, True), False).eq(True)

  def start(self, begin, end, at_end, enabled):
    self.forward.copy_from(end if at_end else begin, enabled)
    self.reverse.copy_from(begin if at_end else end, enabled)
    self.nonempty = choose(enabled, end.focus.present(), self.nonempty)
    phase = Value.constant("end") if at_end else Value.select(end.focus.present(),
               Value.constant("forward"), Value.constant("separator"))
    self.phase = Value.select(enabled, phase, self.phase)

  def read(self, enabled=TRUE):
    front = self.forward.peek_right(conjunction(enabled, self.phase.eq("forward")))
    target = Ref.select(self.phase.eq("forward"), front, self.reverse.focus)
    value = self.circuit.get(target, "input", tuple("ab"), "a")
    data = Value.select(target.present(), value, Value.constant(None))
    return Value.select(self.phase.eq("separator"), Value.constant("#"),
                         Value.select(self.phase.eq("end"), Value.constant(None), data))

  def move(self, direction, enabled):
    phase = self.phase
    forward = conjunction(enabled, phase.eq("forward"))
    separator = conjunction(enabled, phase.eq("separator"))
    reverse = conjunction(enabled, phase.eq("reverse"))
    end = conjunction(enabled, phase.eq("end"))
    if direction == 1:
      self.circuit.require(FALSE, end)
      requested = self.forward.move_right(forward)
      self.reverse.move_left(reverse)
      self.phase = Value.select(conjunction(forward, neg(self.forward.can_right())),
                                Value.constant("separator"), self.phase)
      self.phase = Value.select(separator, Value.select(self.nonempty,
                                Value.constant("reverse"), Value.constant("end")), self.phase)
      self.phase = Value.select(conjunction(reverse, neg(self.reverse.focus.present())),
                                Value.constant("end"), self.phase)
      return self.forward.queue.name, requested
    if direction == -1:
      self.circuit.require(self.nonempty, separator)
      at_first_reverse = neg(self.reverse.can_right())
      self.forward.move_left(disjunction(forward, separator))
      requested = self.reverse.move_right(disjunction(conjunction(reverse, neg(at_first_reverse)),
                                                      conjunction(end, self.nonempty)))
      self.phase = Value.select(separator, Value.constant("forward"), self.phase)
      self.phase = Value.select(conjunction(reverse, at_first_reverse), Value.constant("separator"), self.phase)
      self.phase = Value.select(end, Value.select(self.nonempty,
                                Value.constant("reverse"), Value.constant("separator")), self.phase)
      return self.reverse.queue.name, requested
    raise ValueError("unit direction required")

  def copy_from(self, other, enabled):
    self.forward.copy_from(other.forward, enabled)
    self.reverse.copy_from(other.reverse, enabled)
    self.phase = Value.select(enabled, other.phase, self.phase)
    self.nonempty = choose(enabled, other.nonempty, self.nonempty)

  def finalize(self):
    self.circuit.put(self.phase_key, self.phase)
    self.circuit.put(self.nonempty_key, Value.select(self.nonempty, Value.constant(True), Value.constant(False)))

File: scala/pal/src/main/scala/pal/ScaffoldStreamHeads.scala
package pal

import scala.collection.immutable.VectorMap

import Expr.{TRUE, FALSE}
import ScaffoldCircuit.{conjunction, disjunction, neg, choose}
import Ref.{EMPTY, PREVIOUS}

/** Clonable stream heads using the explicitly scheduled queue kernel.
  *
  * A right move consumes a front cell at an operation boundary and requests
  * three subsequent maintenance instructions if it used a queue. The caller
  * must schedule those before another public operation on that queue.
  *
  * Port of `scaffold_stream_heads.py`. Every `finalize()` is `commit()`.
  */
final class StreamBank(val circuit: Circuit, names: Seq[String]) {
  val cells: StackPool = new StackPool(circuit, Vector("cells" -> 1))
  val queues: QueueRegisters = new QueueRegisters(circuit, names.map(_ + ".in"))
  val heads: VectorMap[String, StreamHead] = VectorMap.from(names.map(name => name -> new StreamHead(this, name)))

  def commit(): Unit = {
    for (head <- heads.values) { head.commit() }
    queues.commit()
  }
}

/** A head over a stream of cells: a focus cell, a left stack, a right stack and an incoming queue. */
final class StreamHead(bank: StreamBank, val name: String) {
  val circuit: Circuit = bank.circuit
  val focusKey: String = name + ".focus"
  var focus: Ref = circuit.getRef(PREVIOUS, focusKey)
  val left: Stack = new Stack(bank.cells, name + ".l", Some("cells"))
  val right: Stack = new Stack(bank.cells, name + ".r", Some("cells"))
  val queue: Queue = bank.queues.queues(name + ".in")

  def canRight(): Expr = disjunction(neg(right.empty()), neg(queue.empty()))

  def peekRight(enabled: Expr = TRUE): Ref = {
    val fromStack = neg(right.empty())
    val (a, _) = right.peek()
    val (b, _) = queue.stacks("F").peek()
    circuit.require(neg(queue.stacks("F").empty()),
                    conjunction(enabled, neg(fromStack), neg(queue.empty())))
    Ref.select(fromStack, a, b)
  }

  /** Move right; returns the guard under which the queue needs maintenance afterwards. */
  def moveRight(enabled: Expr): Expr = {
    circuit.require(canRight(), enabled)
    left.push(focus, enabled = enabled, slot = Some(0))
    val fromStack = neg(right.empty())
    val (a, _) = right.pop(conjunction(enabled, fromStack))
    val maintain = conjunction(enabled, neg(fromStack))
    val b = queue.pop(maintain)
    focus = Ref.select(enabled, Ref.select(fromStack, a, b), focus)
    maintain
  }

  def moveLeft(enabled: Expr): Unit = {
    circuit.require(neg(left.empty()), enabled)
    right.push(focus, enabled = enabled, slot = Some(0))
    val (value, _) = left.pop(enabled)
    focus = Ref.select(enabled, value, focus)
  }

  /** The global end head advances to a newly arrived cell directly. */
  def followArrival(cell: Ref, enabled: Expr): Unit = {
    left.push(focus, enabled = enabled, slot = Some(0))
    focus = Ref.select(enabled, cell, focus)
  }

  def copyFrom(other: StreamHead, enabled: Expr): Unit = {
    focus = Ref.select(enabled, other.focus, focus)
    left.copyFrom(other.left, enabled)
    right.copyFrom(other.right, enabled)
    queue.copyFrom(other.queue, enabled)
  }

  def reset(enabled: Expr): Unit = {
    focus = Ref.select(enabled, EMPTY, focus)
    left.clear(enabled)
    right.clear(enabled)
    queue.clear(enabled)
  }

  def commit(): Unit = {
    left.commit()
    right.commit()
    circuit.putRef(focusKey, focus)
  }
}

/** A head that reads a frozen pattern snapshot, then continues into live text. */
final class PatternTextHead(bank: StreamBank, val name: String) {
  val circuit: Circuit = bank.circuit
  val pattern: StreamHead = bank.heads(name + ".pattern")
  val text: StreamHead = bank.heads(name + ".text")
  val modeKey: String = name + ".text_mode"
  var textMode: Expr = circuit.get(PREVIOUS, modeKey, Vector(false, true), false).eqTo(true)

  def available(): Expr = choose(textMode, text.canRight(), pattern.focus.present())

  /** The symbol under the head, or `None` past the end. */
  def read(enabled: Expr = TRUE): Value[Any] = {
    val target = Ref.select(textMode, text.peekRight(conjunction(enabled, textMode)), pattern.focus)
    val value = circuit.get(target, "input", Vector('a', 'b'), 'a')
    Value.select[Any](target.present(), value, Value.constant(None))
  }

  def start(snapshot: StreamHead, enabled: Expr): Unit = {
    pattern.copyFrom(snapshot, enabled)
    text.reset(enabled)
    textMode = choose(enabled, neg(snapshot.focus.present()), textMode)
  }

  /** Move by one; returns the queue name and the guard requesting its maintenance. */
  def move(direction: Int, enabled: Expr): (String, Expr) = {
    val mode = textMode
    if (direction == 1) {
      pattern.moveLeft(conjunction(enabled, neg(mode)))
      val requested = text.moveRight(conjunction(enabled, mode))
      textMode = disjunction(mode, conjunction(enabled, neg(mode), neg(pattern.focus.present())))
      (text.queue.name, requested)
    } else if (direction == -1) {
      val boundary = neg(text.focus.present())
      text.moveLeft(conjunction(enabled, mode, neg(boundary)))
      val requested = pattern.moveRight(conjunction(enabled, disjunction(neg(mode), boundary)))
      textMode = conjunction(mode, neg(conjunction(enabled, mode, boundary)))
      (pattern.queue.name, requested)
    } else {
      throw new IllegalArgumentException("unit direction required")
    }
  }

  def copyFrom(other: PatternTextHead, enabled: Expr): Unit = {
    pattern.copyFrom(other.pattern, enabled)
    text.copyFrom(other.text, enabled)
    textMode = choose(enabled, other.textMode, textMode)
  }

  def commit(): Unit = {
    circuit.put(modeKey, Value.select(textMode, Value.constant(true), Value.constant(false)))
  }
}

/** A frozen forward or backward view, preserved by head copies. */
final class OrientedHead(bank: StreamBank, val name: String) {
  val circuit: Circuit = bank.circuit
  val cursor: StreamHead = bank.heads(name)
  val modeKey: String = name + ".reverse"
  var reverse: Expr = circuit.get(PREVIOUS, modeKey, Vector(false, true), false).eqTo(true)

  def start(snapshot: StreamHead, reversed: Boolean, enabled: Expr): Unit = {
    cursor.copyFrom(snapshot, enabled)
    reverse = choose(enabled, if (reversed) { TRUE } else { FALSE }, reverse)
  }

  def read(enabled: Expr = TRUE): Value[Any] = {
    val forward = cursor.peekRight(conjunction(enabled, neg(reverse)))
    val target = Ref.select(reverse, cursor.focus, forward)
    val value = circuit.get(target, "input", Vector('a', 'b'), 'a')
    Value.select[Any](target.present(), value, Value.constant(None))
  }

  def move(direction: Int, enabled: Expr): (String, Expr) = {
    if (direction != -1 && direction != 1) { throw new IllegalArgumentException("unit oriented movement required") }
    val right = conjunction(enabled, if (direction == 1) { neg(reverse) } else { reverse })
    val left = conjunction(enabled, if (direction == 1) { reverse } else { neg(reverse) })
    val requested = cursor.moveRight(right)
    cursor.moveLeft(left)
    (cursor.queue.name, requested)
  }

  def copyFrom(other: OrientedHead, enabled: Expr): Unit = {
    cursor.copyFrom(other.cursor, enabled)
    reverse = choose(enabled, other.reverse, reverse)
  }

  def commit(): Unit = {
    circuit.put(modeKey, Value.select(reverse, Value.constant(true), Value.constant(false)))
  }
}

/** A frozen view of u # reverse(u), without building that word. */
final class MirrorHead(bank: StreamBank, val name: String) {
  val circuit: Circuit = bank.circuit
  val forward: StreamHead = bank.heads(name + ".forward")
  val reverse: StreamHead = bank.heads(name + ".reverse")
  val phaseKey: String = name + ".view_phase"
  val nonemptyKey: String = name + ".nonempty"
  var phase: Value[String] = circuit.get(PREVIOUS, phaseKey, Vector("forward", "separator", "reverse", "end"), "separator")
  var nonempty: Expr = circuit.get(PREVIOUS, nonemptyKey, Vector(false, true), false).eqTo(true)

  def start(begin: StreamHead, end: StreamHead, atEnd: Boolean, enabled: Expr): Unit = {
    forward.copyFrom(if (atEnd) { end } else { begin }, enabled)
    reverse.copyFrom(if (atEnd) { begin } else { end }, enabled)
    nonempty = choose(enabled, end.focus.present(), nonempty)
    val next = if (atEnd) {
      Value.constant("end")
    } else {
      Value.select(end.focus.present(), Value.constant("forward"), Value.constant("separator"))
    }
    phase = Value.select(enabled, next, phase)
  }

  /** The viewed symbol: `#` at the separator, `None` at the end. */
  def read(enabled: Expr = TRUE): Value[Any] = {
    val front = forward.peekRight(conjunction(enabled, phase.eqTo("forward")))
    val target = Ref.select(phase.eqTo("forward"), front, reverse.focus)
    val value = circuit.get(target, "input", Vector('a', 'b'), 'a')
    val data = Value.select[Any](target.present(), value, Value.constant(None))
    Value.select[Any](phase.eqTo("separator"), Value.constant('#'),
                      Value.select[Any](phase.eqTo("end"), Value.constant(None), data))
  }

  def move(direction: Int, enabled: Expr): (String, Expr) = {
    val current = phase
    val forwardPhase = conjunction(enabled, current.eqTo("forward"))
    val separator = conjunction(enabled, current.eqTo("separator"))
    val reversePhase = conjunction(enabled, current.eqTo("reverse"))
    val end = conjunction(enabled, current.eqTo("end"))
    if (direction == 1) {
      circuit.require(FALSE, end)
      val requested = forward.moveRight(forwardPhase)
      reverse.moveLeft(reversePhase)
      phase = Value.select(conjunction(forwardPhase, neg(forward.canRight())), Value.constant("separator"), phase)
      phase = Value.select(separator, Value.select(nonempty, Value.constant("reverse"), Value.constant("end")), phase)
      phase = Value.select(conjunction(reversePhase, neg(reverse.focus.present())), Value.constant("end"), phase)
      (forward.queue.name, requested)
    } else if (direction == -1) {
      circuit.require(nonempty, separator)
      val atFirstReverse = neg(reverse.canRight())
      forward.moveLeft(disjunction(forwardPhase, separator))
      val requested = reverse.moveRight(disjunction(conjunction(reversePhase, neg(atFirstReverse)),
                                                    conjunction(end, nonempty)))
      phase = Value.select(separator, Value.constant("forward"), phase)
      phase = Value.select(conjunction(reversePhase, atFirstReverse), Value.constant("separator"), phase)
      phase = Value.select(end, Value.select(nonempty, Value.constant("reverse"), Value.constant("separator")), phase)
      (reverse.queue.name, requested)
    } else {
      throw new IllegalArgumentException("unit direction required")
    }
  }

  def copyFrom(other: MirrorHead, enabled: Expr): Unit = {
    forward.copyFrom(other.forward, enabled)
    reverse.copyFrom(other.reverse, enabled)
    phase = Value.select(enabled, other.phase, phase)
    nonempty = choose(enabled, other.nonempty, nonempty)
  }

  def commit(): Unit = {
    circuit.put(phaseKey, phase)
    circuit.put(nonemptyKey, Value.select(nonempty, Value.constant(true), Value.constant(false)))
  }
}
